drivers/flash/spi_nor.c

```diff
@@ -110,6 +185,88 @@ static int spi_nor_access(const struct device *const dev,
 #define spi_nor_cmd_addr_write(dev, opcode, addr, src, length) \
 	spi_nor_access(dev, opcode, true, addr, (void *)src, length, true)
 
+static int enter_dpd(const struct device *const dev)
+{
+	int ret = 0;
+
+	if (IS_ENABLED(DT_INST_0_JEDEC_SPI_NOR_HAS_DPD)) {
+		ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_DPD);
+		if (ret == 0) {
+			record_entered_dpd(dev);
+		}
+	}
+	return ret;
+}
+
+static int exit_dpd(const struct device *const dev)
+{
+	int ret = 0;
+
+	if (IS_ENABLED(DT_INST_0_JEDEC_SPI_NOR_HAS_DPD)) {
+		delay_until_exit_dpd_ok(dev);
+
+#ifdef DT_INST_0_JEDEC_SPI_NOR_DPD_WAKEUP_SEQUENCE
+		/* Assert CSn and wait for tCRDP.
+		 *
+		 * Unfortunately the SPI API doesn't allow us to
+		 * control CSn so fake it by writing a known-supported
+		 * single-byte command, hoping that'll hold the assert
+		 * long enough.  This is highly likely, since the
+		 * duration is usually less than two SPI clock cycles.
+		 */
+		ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_RDID);
+
+		/* Deassert CSn and wait for tRDP */
+		k_sleep(K_MSEC(T_RDP_MS));
+#else /* DPD_WAKEUP_SEQUENCE */
+		ret = spi_nor_cmd_write(dev, SPI_NOR_CMD_RDPD);
+
+		if (ret == 0) {
+#ifdef DT_INST_0_JEDEC_SPI_NOR_T_EXIT_DPD
+			k_sleep(K_MSEC(T_RES1_MS));
+#endif /* T_EXIT_DPD */
+		}
+#endif /* DPD_WAKEUP_SEQUENCE */
+	}
+	return ret;
+}
+
+/* Everything necessary to acquire owning access to the device.
+ *
+ * This means taking the lock and, if necessary, waking the device
+ * from deep power-down mode.
+ */
+static void acquire_device(struct device *dev)
+{
+#ifdef CONFIG_MULTITHREADING
+	struct spi_nor_data *const driver_data = dev->driver_data;
+
+	k_sem_take(&driver_data->sem, K_FOREVER);
```

Comment:
sem attribute seem to be always present, so you could use IS_ENABLED(CONFIG_MULTITHREAD) here